modules/yarpc/middleware.go

```diff
@@ -34,33 +36,57 @@ import (
 
 const _panicResponse = "Server Error"
 
-type contextInboundMiddleware struct {
-	statsClient *statsClient
+var (
+	mu sync.Mutex
+)
+
+// TransportUnaryMiddleware keeps all the decorator layers defined in the configuration
+type TransportUnaryMiddleware struct {
+	procedures map[string][]decorator.Decorator
+	decorators map[string]transport.UnaryHandler
 }
 
-func (f contextInboundMiddleware) Handle(
+// Handle all decorators for Unary transport
+func (l TransportUnaryMiddleware) Handle(
 	ctx context.Context,
 	req *transport.Request,
 	resw transport.ResponseWriter,
 	handler transport.UnaryHandler,
 ) error {
-	stopwatch := f.statsClient.RPCHandleTimer().
-		Tagged(map[string]string{_tagProcedure: req.Procedure}).
-		Timer(req.Procedure).
-		Start()
-	defer stopwatch.Stop()
+	if _, ok := l.procedures[req.Procedure]; !ok {
+		return handler.Handle(ctx, req, resw)
+	}
 
-	return handler.Handle(ctx, req, resw)
+	mu.Lock()
```

Comment:
With the current implementation, we will only lock upon the first call to `Handle`. When Anup and I explored this earlier, it seemed that the initialization needed to happen here (given the current state of FX+YARPC). There may be a way to refactor this in the future, but it will potentially require changes in both repositories.